### ISSUE TITLE
[raspberry] Poor video performance (low fps, stuck picture loading)

### ISSUE THREAD

**Issue by pthoelken** [FreeRDP/FreeRDP#5917]
Hello Members, 

since today we're use an Raspberry Pi 4 with xfreerdp for remote desktop connection. The problem is the video quality and the connection speed. When I open up a video file, YouTube or simple a GIF file, it is picture by picture. The frame rate is never more then 5-8 fps. If it's needed I can show you my connection parameters tomorrow (today it's end of working time). 

We've tested the wifi rdp connection with another, windows device and it was very good and fast. 

Anybody knows an idea about this problem?

Thanks Patrick

**Comment by akallabeth** [comment]
Well, you could start with providing some version and command line information ;)
Without it is quite hard to guess

**Comment by pthoelken** [comment]
Absolutely understand this. I will paste my parameters and version tomorrow morning. Thanks in advance.

**Comment by pthoelken** [comment]
We're use xfreerdp Version **2.0.0-dev5 (2693389a+debian)** and with the following command to connect. 

`xfreerdp /v:tsfarm.foo.bar /cert-ignore +credentials-delegation /f /rfx /gfx:avc444 /u:user01@foo.bar /p:PasswordZY +clipboard`

**Comment by akallabeth** [comment]
@pthoelken try adding /network:auto and if that does not work well only /gfx or /rfx
The first one activates bandwidth detection, the second one lrss cpu hungry codecs.
You should verify what is actually used by checking with /log-level:debug as the server has the last word in this regard (the client just tells what is supported)

**Comment by pthoelken** [comment]
Okay. Thanks. Where is the log located? What do you preffere /gfx or /rfx? And with :avc444 after the parameter or blank?

**Comment by akallabeth** [comment]
There is no log, just terminal output. Leave out the avc444 for the second one.

**Comment by pthoelken** [comment]
Okay. With /network:auto only the quality is much bad then before. I will try it as the follow: 

- /network:auto /gfx:avc444 == bad quality, still have only 2-3 fps
- /network:auto /gfx:avc444 /rfx == bad quality, still have only 2-3 fps
- /network:auto == bad quality, still have only 2-3 fps

**Comment by akallabeth** [comment]
Well, the test cases with only /rfx and only /gfx are missing ;)

**Comment by pthoelken** [comment]
I've updated the post before, with the test results. So, down below pasted the log and terminal results. 

**This cames from terminal after terminate the command:**
[07:11:13:489] [954:964] [ERROR][com.freerdp.channels.drdynvc.client] - ChannelId 10 not found!
[07:11:14:554] [954:964] [ERROR][com.freerdp.channels.drdynvc.client] - ChannelId 10 not found!
[h264 @ 0xa7b1df10] slice type 32 too large at 95
[h264 @ 0xa7b1df10] decode_slice_header error
[h264 @ 0xa7b1df10] slice type 32 too large at -1
[h264 @ 0xa7b1df10] decode_slice_header error
[h264 @ 0xa7b1df10] slice type 32 too large at -1
[h264 @ 0xa7b1df10] decode_slice_header error
[h264 @ 0xa7b1df10] slice type 32 too large at 20287
[h264 @ 0xa7b1df10] decode_slice_header error

**This cames from the log (I've started xfreerdp with "> /home/pi/log.log"**

[debuglog.log](https://github.com/FreeRDP/FreeRDP/files/4259789/debuglog.log)

**Comment by akallabeth** [comment]
@pthoelken ok, looks like you have a really interesting server side ;)
windows 8.1 (or the server equivalent) based?
Better use something like `/rfx /gfx` or just `/rfx` in this case, the H264 implementation there was never really working properly.

**Comment by pthoelken** [comment]
It is Windows Server 2012 R2 Datacenter Version :) 

So you mean I should use it like this? 

`xfreerdp /v:tsfarm.foo.bar /cert-ignore +credentials-delegation /f /rfx /gfx /u:user01@foo.bar /p:PasswordZY`

What can I make if it's not working better?

**Comment by pthoelken** [comment]
Other thing is, that the keyboard layout forwarding not work. Still have US layout. I need the german layout.

I use /kbd:0x407 - but when i'm connected I have still the US layout.

**Comment by akallabeth** [comment]
@pthoelken please do not discuss other stuff in issues, for such things we have a mailing list and IRC.
But see #5707 (only available with nightlies) 
Keyboard language announcement is a bit of a pain with RDP

**Comment by akallabeth** [comment]
@pthoelken try checking out master and build yourself. (and compile against OpenH264 to have a faster H264 decoder)

**Comment by akallabeth** [comment]
@pthoelken I´ve forgotten to ask, we´re talking about a 1080p monitor here and not 4k?
The former should work quite fine, 4k though is a bit too much for the PI CPU (we have no GPU decoding support currently)

**Comment by pthoelken** [comment]
Hi, yes. We're talking about 1080p. I will check out and build is in the next days. Thanks.

**Comment by akallabeth** [comment]
@pthoelken ok, then you should check your graphics driver, there are great differences there too.
I have a PI3 and can run with ~15FPS so the PI4 should definitely be faster

**Comment by AndersSahlin** [comment]
Have you tried changing to the legacy GL driver in raspi-config? According to this thread the new video driver might have issues with RDP:

https://www.raspberrypi.org/forums/viewtopic.php?f=63&t=249628&sid=fead2bc6712197de3f12bf236b64d3a2

**Comment by henk717** [comment]
Having this as well on my Pi4, comparing this to other applications like the Steam Home Streaming app or hardware accelerated H264 playback the performance is really lackluster. However, compared to H264 software playback it seems to be similar.
Perhaps some of the developers can shine a light on this but does xfreerdp support hardware decoding at all? I tried compiling it with ffmpeg instead of the version in the raspbian testing repository but this made no difference.

I did try the legacy GL driver, but on my Pi4 image that resulted in the xorg server not starting at all this might be Pi3 only.
And its not a matter of parameters for me since its related to the decode and/or render speed of the H264 stream. You see the frames being drawn in from the top and depending on resolution (I tested on 720P and 1080P) the performance differs but is sub par compared to using xfreerdp on other platforms.

**Comment by hardening** [comment]
What does say vainfo ?

**Comment by akallabeth** [comment]
@henk717 software decoding only. with `FFMPEG` you might be able to offload (some) decoding work, but RDP uses some customized codecs/extensions that have no hardware acceleration available.
If you already recompiled `xfreerdp` ensure that you use the highest compiler optimisations available, that does improve performance quite a bit (but not enough for (completely) smooth video

**Comment by henk717** [comment]
After much playing around and compiling my own with maximum optimizations it is the V3D driver of the Pi that is causing this.
So people enabling the legacy driver bypass this issue resulting in a big speed boost with reduced screen detection and multimonitor support. I have found a better way that also seems to be applied to the Thinlinx image, compiling the Linux kernel yourself but disabling the V3D driver. To do this follow the raspberry pi kernel compilation tutorial and use the interactive customization menu. Navigate towards the graphics driver and uncheck the M next to V3D. Once you compile and install this kernel you will have the same speed benefit as legacy mode, but your monitor resolution should still be detected.

I have filed a bug report in the raspberrypi kernel github to notify them that V3D is causing this slowdown issue in hopes a graphics driver developer can take a look at it further. You can find the bug report here : https://github.com/raspberrypi/linux/issues/4353

**Comment by kishorv06** [comment]
> 

@akallabeth  Can you please elaborate what you meant by `software decoding only` ? Is it a limitation only on the Raspberry Pi or is it a limitation of RDP due to those fancy extensions ? Or is it not yet implemented on FreeRDP ?

**Comment by akallabeth** [comment]
@kishorv06 `FreeRDP` currently does not utilize `vulkan` or any other method to move decoding load to the `GPU` (raspberry is not very good there either)
Most of the load a RDP connection creates are:
* SSL related encoding/decoding
* Video decoding
* Video stream merging (they use some very specialized codecs that need intensive post processing)
* Drawing

So, while the `SSL` related part is taken care of by a proper `OpenSSL` configuration and `Video decoding` can partly be accelerated by using `FFMPEG` (h264 mode only) the `Video stream merging` can only be done properly by using something like `vulkan shaders`

**Comment by Nabla128k** [comment]
Hi. Since this issue has not been resolved yet on the side of v3d driver and now we have 64bit build of Raspberry OS, which desktop performance seems to be havily dependent on v3d, now freerdp is not in fact usable on Raspberry OS 64bit. Is there any chance to resolve the issue on the side of FreeRDP? Thanks.

**Comment by lollo0296** [comment]
@Nabla128k I can confirm your statement. 
FreeRDP is not really usable on RPi4, neither 32 nor 64 bit RaspberryOS. About 5 years ago we started using RPis in our company as a thinclient for RDP into our Windows Server RDS-Farm. We have a mixture of 3Bs and 3B+s and it worked great with `Server 2008R2`. Now we have `Server 2022` and the 3B / 3B+ still works like a charm.

I tried for over 2 years to make it work on 4B, tried all the existing versions of FreeRDP, including GitHub master and other branches,tried all possible combinations of `cmake` build parameters, `ffmpeg` versions, `openh264` versions and command line arguments (`/sfx`, `/gdx`, /`gdx:xxxxx` and so on...).

It never worked as well as with Raspberry Pi 3. The best experience actually comes with standard builds (no fancy parameters or experimentations) and `/gdx` only. But it's still not good enough for daily business usage.

Initially I was only interested in RPi4 because of `/multimon`feature, but now RPi3s are not available anymore and, as some of existing one need to be replaced and the number of workplaces grows, we have no choice other than buy 4B (good luck finding one), if we want to stick with the RPi thinclient solution.

Is there something we can actively do to help solving this problem?

**Comment by Nabla128k** [comment]
@lollo0296 Temporary solution on 4B was to use legacy GL driver (not fkms) or blacklist v3d driver. It can be used with dual monitor and the performance is quite good (I use it every day now for some months). But, with new version of Raspberry Pi OS (11) it's just more problematic. Because there's no "legacy" driver, just kms. You can disable v3d but desktop is not usable (mutter). You can switch from muter back to openbox, but I personally don't want so deep tweaking of "stock" OS (other thing is probably performance drop of chrome, vlc...).

So please, can someone here help, or what can we do to resolve this situation?

**Comment by akallabeth** [comment]
@lollo0296 bad news then, `xfreerdp` did not change the way it does drawing for a decade, you can stop experimenting with different versions.
This looks like the newer hardware (or the driver) is simply (a lot) slower on drawing plain bitmap graphics? (might be the bandwidth between CPU/memory and GPU?)

Could you try with different clients?
* Remmina is based on FreeRDP and uses GTK.
* https://flathub.org/apps/details/com.thincast.client (a RDP client based on FreeRDP which draws using QT/vulkan) 

and check if performance is better/worse?

**Comment by Nabla128k** [comment]
> Could you try with different clients?
> 
> * Remmina is using FreeRDP based RDP and uses GTK for drawing.
> * https://flathub.org/apps/details/com.thincast.client (a RDP client based on FreeRDP which draws using QT/vulkan)
> 
> and check if performance is better/worse?

Tried "Thincast" and the result is ... quite good (better then with xfreerdp) but I can't manage to work it with vulkan. I've installed mesa-vulkan-drivers and I believe vulkan is working on my machine (glxgear, vkcube demos). But thincast client (Thincast Remote Desktop Client version 1.0.268 (stable tcc-stable-1.0.268)) says:

`vkDebug: Loader Message: 0: setupLoaderTermPhysDevs:  Failed to detect any valid GPUs in the current config
vkDebug: Loader Message: 0: setupLoaderTrampPhysDevs:  Failed during dispatch call of 'vkEnumeratePhysicalDevices' to lower layers or loader to get count.
[20:44:33:644] [T:    2] [error] No Vulkan physical device available (-3).`

Tried also Remmina and it's slow the same way as xfreerdp

(just for future visitors)

**Comment by henk717** [comment]
For my own tests I used raw xFreeRDP, no clients. Anything using xFreeRDP will likely give the same results, i believe thincast has a different solution. Citrix should also behave better for example.

**Comment by akallabeth** [comment]
@Nabla128k that sounds good (that it is faster) and strengthens the assumption that it is driver dependent (it uses qt instead of x11 for drawing, which utilize different optimizations/drawing primitives).
It is also normal that there is no vulkan device detected (raspi vulkan drivers are nowhere complete and lack sufficient support for what is required here, that was the state at least the last time I´ve tried)

**Comment by SupraGuy** [comment]
Same issue here, Raspberry Pi 4B. Is fine on the older server, but on the new server (server 2019) it is unusable.

I will look into thincast, but would have liked to have found the issue fixed...

**Comment by mnordstr** [comment]
So after fighting with this for some time I've found a solution that works for us. Unfortunately it requires disabling the v3d driver so not a very long term solution but Raspberry Pi will hopefully make the new driver better at some point.

So first step is disabling the driver in /boot/config.txt, just commenting out the line dtoverlay=vc4-kms-v3d is enough.

That will make the general UI sluggish but a simple solution is to enable the VNC server, that will actually change the window manager to openbox (or doing that manually). After that (reboot) the UI is normal and the RDP connection works like before.

This is what @Nabla128k suggested above, but IMO not that bad as we already used the VNC server so the only real change was to disable the v3d driver... Hope this saves someone some gray hairs :)

**Comment by jankkm** [comment]
> So after fighting with this for some time I've found a solution that works for us. Unfortunately it requires disabling the v3d driver so not a very long term solution but Raspberry Pi will hopefully make the new driver better at some point.
> 
> So first step is disabling the driver in /boot/config.txt, just commenting out the line dtoverlay=vc4-kms-v3d is enough.
> 
> That will make the general UI sluggish but a simple solution is to enable the VNC server, that will actually change the window manager to openbox (or doing that manually). After that (reboot) the UI is normal and the RDP connection works like before.
> 
> This is what @Nabla128k suggested above, but IMO not that bad as we already used the VNC server so the only real change was to disable the v3d driver... Hope this saves someone some gray hairs :)

Can you tell me what you are working with to begin with?
I have the current Raspi OS light with the lxde desktop installed (which to my knowledge uses openbox) and when commenting 'dtoverlay=vc4-kms-v3d' it doesn't even show the desktop. Installing the realvnc-server via raspi-config does not help either. It'd be great if you could give me a hint on how to get it working. Thanks!

**Comment by mnordstr** [comment]
> 
> Can you tell me what you are working with to begin with? I have the current Raspi OS light with the lxde desktop installed (which to my knowledge uses openbox) and when commenting 'dtoverlay=vc4-kms-v3d' it doesn't even show the desktop. Installing the realvnc-server via raspi-config does not help either. It'd be great if you could give me a hint on how to get it working. Thanks!

We use Raspberry OS (64-bit). Not the light version but not the full either. I think we tested on 32bit also, so basically the default option for the Raspberry Pi Imager. Double checked our Ansible script and this is the only relevant thing for a clean os:

```yaml
- name: Disable new GL driver
  lineinfile:
    path: /boot/config.txt
    regexp: '^dtoverlay=vc4-kms-v3d'
    line: "#dtoverlay=vc4-kms-v3d"
  notify: "restart host"

- name: Install software
  apt:
    name:
      - freerdp2-x11
      - cups
      - realvnc-vnc-server
    state: present
    update_cache: yes
    cache_valid_time: 86400
  become: yes

- name: Enable service vncserver-x11-serviced and ensure it is not masked
  ansible.builtin.systemd:
    name: vncserver-x11-serviced
    enabled: yes
    masked: no
  become: yes
```

So remove (comment out) dtoverlay=vc4-kms-v3d in /boot/config.txt and make sure the realvnc-server is enabled.

If it doesn't show the desktop it might be another issue. Had a similar problem with some devices using a DVI monitor (with HDMI-DVI cable). Had to put the hdmi mode manually in /boot/config.txt after disabling the v3d driver.

[config.txt docs](https://www.raspberrypi.com/documentation/computers/config_txt.html)

Adding this fixed it in our case for a 1920x1080 DVI monitor:

hdmi_group=2
hdmi_mode=82

**Comment by jankkm** [comment]
@mnordstr Thank you, that helped! This makes it quite similar to the performance of our RP3 setup.
